tools/git-change-exec/main.go

```diff
@@ -0,0 +1,307 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/plumbing/storer"
+	"github.com/spf13/cobra"
+)
+
+var dryRun = true
+var debug = false
+
+type gitChangeExec struct {
+	actionsToCheck []action
+	actionDos      map[action]struct{}
+	gitPath        string
+	g              *git.Repository
+	visitedPaths   map[string]struct{}
+}
+
+func debugLog(fmt string, args ...any) {
+	if debug {
+		log.Printf(fmt, args...)
+	}
+}
+
+func newGitChangeExec() gitChangeExec {
+	return gitChangeExec{
+		actionDos:      map[action]struct{}{},
+		actionsToCheck: []action{},
+		visitedPaths:   map[string]struct{}{},
+	}
+}
+
+func main() {
+
+	actionCategories := []string{}
+
+	for c := range actions {
+		actionCategories = append(actionCategories, c)
+	}
+
+	rootCmd := cobra.Command{
+		Args: cobra.MinimumNArgs(1),
+		Use:  strings.Join(actionCategories, "|"),
+		Run: func(cmd *cobra.Command, args []string) {
+			gce := newGitChangeExec()
+
+			for _, cat := range args {
+				if len(actions[cat]) == 0 {
+					log.Fatalf("could not find actions for %s", cat)
+				}
+
+				gce.actionsToCheck = append(gce.actionsToCheck, actions[cat]...)
+			}
+
+			currentPath, err := os.Getwd()
+			if err != nil {
+				log.Fatalf("getting current working directory: %v", err)
+			}
+
+			defer func() {
+				err := os.Chdir(currentPath)
+				if err != nil {
+					log.Printf("could not change back to previous dir %s: %v", currentPath, err)
+				}
+			}()
+
+			gce.g, err = git.PlainOpenWithOptions("./", &git.PlainOpenOptions{DetectDotGit: true})
+			if err != nil {
+				log.Fatalf("open git path %s failed: %v", gce.gitPath, err)
+			}
+
+			wt, err := gce.g.Worktree()
+			if err != nil {
+				log.Fatalf("could not determine worktree: %v", err)
+			}
+			gce.gitPath = wt.Filesystem.Root()
+
+			err = os.Chdir(gce.gitPath)
+			if err != nil {
+				log.Fatalf("could not change to %s: %v", gce.gitPath, err)
+			}
+
+			gce.fetchOrigin()
+
+			gce.collectActionsGitTree()
+			gce.collectDirtyGitTree()
```

Comment:
I need to read the function code to understand the purpose of these calls or the difference between them. I guess they define the actions to be run, but the names could be more clear.

Reply:
For the second function, I would make it optional, depending on the flag. If I prepare a PR, I still can have some modified files that will not go into the commits: ./conf/*, for example - and I don't need to check them.

Reply:
I would prefer to make it opt-out with `--ignore-dirty` or perhaps:
* `git-change-exec run test` to run as right now
* `git-change-exec run-branch test` to only consider `collectActionsGitTree()`

But I think the behavior could then be a bit unexpected, f.e. if you change two files in pillar, then no action will be triggered, now you `git add` one of those files, then the tests for pillar will be triggered, but the tests include the second file that has not been added yet.